src/appleseed/foundation/image/oiioimagefilewriter.cpp

```diff
@@ -0,0 +1,515 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "oiioimagefilewriter.h"
+
+// appleseed.foundation headers.
+#include "foundation/core/exceptions/exceptionioerror.h"
+#include "foundation/image/icanvas.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/foreach.h"
+#include "foundation/utility/iostreamop.h"
+#include "foundation/utility/string.h"
+
+// Boost headers.
+#include "boost/filesystem/path.hpp"
+
+// Standard headers.
+#include <algorithm>
+#include <memory>
+#include <stdexcept>
+
+namespace foundation
+{
+
+struct OIIOImageFileWriter::OIIOImages
+{
+    std::vector<const ICanvas*>     m_canvas;
+    std::vector<OIIO::ImageSpec>    m_spec;
+};
+
+OIIOImageFileWriter::OIIOImageFileWriter(const char* filename) :
+    m_images{ new OIIOImages{} }
+{
+    if (filename == nullptr)
+        throw ExceptionIOError("Valid filename must be provide!");
+
+    m_filename = filename;
+
+    m_writer = OIIO::ImageOutput::create(m_filename);
+    if (m_writer == nullptr)
+    {
+        const std::string msg = OpenImageIO::geterror();
+        throw ExceptionIOError(msg.c_str());
+    }
+}
+
+OIIOImageFileWriter::~OIIOImageFileWriter()
+{
+    // Destroy the ImageOutput stucture.
+    OIIO::ImageOutput::destroy(m_writer);
+
+    m_writer = nullptr;
+    m_filename = nullptr;
+
+    m_images->m_canvas.clear();
+    m_images->m_spec.clear();
+
+    delete m_images;
+}
+
+size_t OIIOImageFileWriter::get_image_count(void) const
+{
+    return m_images->m_canvas.size();
+}
+
+void OIIOImageFileWriter::append_image(const ICanvas* image)
+{
+    if (image == nullptr)
+        throw ExceptionIOError("Valid image must be provide!");
+
+    m_images->m_canvas.push_back(image);
+    m_images->m_spec.push_back(OIIO::ImageSpec{});
+
+    set_image_spec();
+}
+
+OIIO::TypeDesc convert_pixel_format(PixelFormat format)
+{
+    switch (format)
+    {
+    case PixelFormatUInt8:
+        return OIIO::TypeDesc::UINT8;
+
+    case PixelFormatUInt16:
+        return OIIO::TypeDesc::UINT16;
+
+    case PixelFormatUInt32:
+        return OIIO::TypeDesc::UINT32;
+
+    case PixelFormatHalf:
+        return OIIO::TypeDesc::HALF;
+
+    case PixelFormatFloat:
+        return OIIO::TypeDesc::FLOAT;
+
+    case PixelFormatDouble:
+        return OIIO::TypeDesc::DOUBLE;
+
+    default:
+        return OIIO::TypeDesc::UNKNOWN;
+    }
+}
+
+void OIIOImageFileWriter::set_image_output_format(const PixelFormat output_pixel_format)
+{
+    if (get_image_count() == 0)
```

Comment:
I think an assertion would be more appropriate here:

    assert(!m_images->m_spec.empty());

Fix everywhere.